core/distributed/matrix.cpp

```diff
@@ -0,0 +1,440 @@
+    } else {
+        local_mtx_->apply(local_alpha, dense_b->get_local_vector(), local_beta,
+                          local_x.get());
+    }
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(const Matrix& other)
+    : EnableLinOp<Matrix<value_type, local_index_type,
+                         global_index_type>>{other.get_executor()},
+      DistributedBase{other.get_communicator()}
+{
+    *this = other;
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    Matrix&& other) noexcept
+    : EnableLinOp<Matrix<value_type, local_index_type,
+                         global_index_type>>{other.get_executor()},
+      DistributedBase{other.get_communicator()}
+{
+    *this = std::move(other);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>&
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::operator=(
+    const Matrix& other)
+{
+    if (this != &other) {
+        GKO_ASSERT_EQ(other.get_communicator().size(),
+                      this->get_communicator().size());
+        this->set_size(other.get_size());
+        local_mtx_->copy_from(other.local_mtx_.get());
+        non_local_mtx_->copy_from(other.non_local_mtx_.get());
+        gather_idxs_ = other.gather_idxs_;
+        send_offsets_ = other.send_offsets_;
+        recv_offsets_ = other.recv_offsets_;
+        recv_sizes_ = other.recv_sizes_;
+        send_sizes_ = other.send_sizes_;
+        recv_sizes_ = other.recv_sizes_;
+        non_local_to_global_ = other.non_local_to_global_;
+        one_scalar_.init(this->get_executor(), dim<2>{1, 1});
+        one_scalar_->fill(one<value_type>());
+    }
+    return *this;
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>&
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::operator=(Matrix&& other)
+{
+    if (this != &other) {
+        GKO_ASSERT_EQ(other.get_communicator().size(),
+                      this->get_communicator().size());
+        this->set_size(other.get_size());
+        other.set_size({});
+        local_mtx_->move_from(other.local_mtx_.get());
+        non_local_mtx_->move_from(other.non_local_mtx_.get());
+        gather_idxs_ = std::move(other.gather_idxs_);
+        send_offsets_ = std::move(other.send_offsets_);
+        recv_offsets_ = std::move(other.recv_offsets_);
+        recv_sizes_ = std::move(other.recv_sizes_);
+        send_sizes_ = std::move(other.send_sizes_);
+        recv_sizes_ = std::move(other.recv_sizes_);
+        non_local_to_global_ = std::move(other.non_local_to_global_);
+        one_scalar_.init(this->get_executor(), dim<2>{1, 1});
+        one_scalar_->fill(one<value_type>());
```

Comment:
You can also probably just move from `other.one_scalar_` ?

Reply:
I think filling it explicitly here is the safer choice.